fqdn / policy: plumb mapping of FQDNSelector --> set of IPs into SelectorCache

ianvernon opened the pull request:
This is the minimally-viable solution to plumb the mapping of FQDNSelector --> set of IPs into the new `SelectorCache` construct, which will be used when computing policy for a given identity being consumed by an Endpoint that is running locally.

DNS policy is still transmitted through Cilium via adding / deleting rules from the policy repository with these changes. However, the mapping of FQDNs --> IPs is plumbed as well. This is done so we can start using the `SelectorCache` for all `*Selector` --> identity lookups while computing policy.

Note that removing of the policy rule update / add / remove logic from the FQDN side is out of scope of this PR. The policy computation layer is insulated from this, so this can be followed-up later. 

I also did not handle the case where we fetch the IPs that may map to the given FQDN that is provided in a new rule with  `toFQDNs` rules (`PrepareToFQDNsRules`). Given that we will plumb upon DNS lookup, I figured this would be OK for now - @raybejjani would be interested in your feedback here, though.  

TODO in follow-up PRs:

- [ ] lifecycle management of Selectors in the SelectorCache - need to iterate through rules that are being deleted from policy and remove `FQDNSelectors` from the `SelectorCache`
- [ ] removal of CIDR rule generation

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7879)
<!-- Reviewable:end -->

Commit:
policy: add FQDNSelector handling to SelectorCache

Add the plumbing for relating sets of identities to a given FQDNSelector from policy.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy API: add String() function for FQDNSelector

Similar to the EndpointSelector, this allows for storing the string
representation of an FQDNSelector as a map key in the SelectorCache.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
ipcache: return set of allocated identities from AllocateCIDRs

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
identity: add String() function for Identity

This returns the numeric identity of the Identity as a string.

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -199,9 +244,167 @@ func (l *labelIdentitySelector) updateSelections() {
+	// getting the CIDR identities which correspond to this FQDNSelector. This
+	// is the primary difference here between FQDNSelector and IdentitySelector.
+	fqdnSel.updateSelections()
+	//fmt.Printf("key: %s, added: %v, deleted: %v\n", fqdnKey, added, deleted)
```
remove this

ianvernon commented on pkg/policy/cidr.go:
```diff
@@ -68,3 +68,26 @@ func GetCIDRPrefixes(rules api.Rules) []*net.IPNet {
 }
+
+// GetCIDRPrefixesFromIPs returns all of the ips as a slice of *net.IPNet.
+func GetCIDRPrefixesFromIPs(ips []net.IP) []*net.IPNet {
```
This could probably be put into `pkg/ip`.

joestringer replied:
Yes please, this looks like exactly the kind of function that can end up existing in several places in the codebase if we don't put it somewhere generic or don't go looking for it :-)

ianvernon commented on pkg/fqdn/rulegen.go:
```diff
@@ -274,25 +284,24 @@ func (gen *RuleGen) UpdateGenerateDNS(lookupTime time.Time, updatedDNSIPs map[st
 // names in namesToRegen. These names are FQDNs and toFQDNs.matchPatterns or
 // matchNames that match them will cause these rules to regenerate.
 func (gen *RuleGen) ForceGenerateDNS(namesToRegen []string) error {
+	log.Infof("ForceGenerateDNS for namesToRegen: %s", namesToRegen)
```
Remove this.

ianvernon commented on pkg/fqdn/rulegen.go:
```diff
@@ -274,25 +284,24 @@ func (gen *RuleGen) UpdateGenerateDNS(lookupTime time.Time, updatedDNSIPs map[st
 	}
 
-	// no new rules to add, do not call AddGeneratedRules below
-	if len(generatedRules) == 0 {
```
This is a notable change cc: @raybejjani because we want to always update the SelectorCache. Please shout if you think I need to do more work here. I elected to not do so because CIDR rule generation is going to be removed soon anyway.

raybejjani replied:
I don't remember why this ended up like this. It probably was to avoid waiting on locks since PolicyAdd no-oped on no rules (I think). I think this is totally fine. We can always guard against it elsewhere, which may be better anyway.

An alternative would have been to maintain two callbacks, one for selectors that is called unconditionally, and another for the older datapath with the length logic. We're probably too far along for that, though.

ianvernon commented on daemon/fqdn.go:
```diff
@@ -73,11 +76,88 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
+				log.WithFields(logrus.Fields{
+					"fqdnSelectorString": selectorsWithoutIPs[k],
+				}).Debug("removing all identities from FQDN selector")
+				policy.UpdateFQDNSelector(selectorsWithoutIPs[k], []identity.NumericIdentity{})
```
We may need additional synchronization around this. Should we use `EventQueue` or `pkg/serializer` ? I bet we are relying upon the synchronizing of the policy repository via `EventQueue` right now to ensure that DNS updates are handled in proper order.

joestringer replied:
Hmm, good point. My initial thought was something more like a `Trigger`. Not sure exactly how you want to coalesce frequent updates of this kind such that the SelectorCache issues periodic batches of changes, perhaps the calculation of the diff between the old set of identities selected by the FQDNSelector and the new set of identities selected is determined in the Trigger function?

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -125,6 +126,50 @@ type labelIdentitySelector struct {
+// that case GetSelections() will return either the old or new version
+// of the selections. If the old version is returned, the user is
+// guaranteed to receive a notification including the update.
+func (l *fqdnSelector) GetSelections() []identity.NumericIdentity {
```
Looks to me like there's a bunch of these functions which are identical to the existing `labelIdentitySelector`, so some minor refactoring could allow us to compose this together.

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -199,9 +244,167 @@ func (l *labelIdentitySelector) updateSelections() {
+	var added, deleted []identity.NumericIdentity
+
+	_ = len(identities)
+	_ = len(fqdnSel.cachedSelections)
```
Unused statements.

aanm replied:
+1

joestringer commented on pkg/identity/identity.go:
```diff
@@ -109,6 +109,11 @@ func (id *Identity) StringID() string {
 }
 
+// StringID returns the identity identifier as string
+func (id *Identity) String() string {
```
Is this meaningfully distinct from `StringID()`? Also, the comment refers to `StringID()`, not `String()`.

ianvernon replied:
It makes `*Identity` implement the `Stringer` interface, which allows for more readable debug logs when using `%s` for the identity. I can add other fields as well, including labels.

joestringer replied:
This response satisfies the `Stringer` interface. ;)

(In all seriousness, for CIDR identities the labels may become quite cumbersome, not sure whether it would actually make it easier if more fields are included)

ianvernon replied:
I'm ambivalent. If we find that we need more information from logging, we can always add it later if need be?

joestringer commented on daemon/fqdn.go:
```diff
@@ -73,11 +76,88 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
 		Cache:          fqdn.NewDNSCache(option.Config.ToFQDNsMinTTL),
 		LookupDNSNames: fqdn.DNSLookupDefaultResolver,
-		AddGeneratedRules: func(generatedRules []*policyApi.Rule) error {
+		AddGeneratedRulesAndUpdateSelectors: func(generatedRules []*policyApi.Rule, selectorWithIPsToUpdate map[string][]net.IP, selectorsWithoutIPs []string) error {
```
I'm not sure what your longer-term plans are for this, but please consider splitting this out into smaller functions. This is not unit-testable, and it's is so long that it's easy to forget you're still just initializing a struct by the end of it. (The existing code also suffers from this..)

aanm replied:
+1

aanm replied:
I must say, please write this function somewhere else instead directly here.

joestringer commented on daemon/fqdn.go:
```diff
@@ -73,11 +76,88 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
+		AddGeneratedRulesAndUpdateSelectors: func(generatedRules []*policyApi.Rule, selectorWithIPsToUpdate map[string][]net.IP, selectorsWithoutIPs []string) error {
+
+			// Prefix is always a /32. Maybe we can just add a /32 to the
+			// prefixes or /128 for IPV4, IPv6?
```
Already done, please see `createPrefixLengthCounter()`:

https://github.com/cilium/cilium/blob/9e81aff817bf7a29c7a4782ddbdedc33bc3bace2/daemon/daemon.go#L867-L890

This should also simplify the error paths below.

joestringer commented on daemon/fqdn.go:
```diff
@@ -73,11 +76,88 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
+				return err
+			}
+
+			if newPrefixLengths && !bpfIPCache.BackedByLPM() {
```
Not necessary per the above. we always support full host prefixes.

joestringer commented on daemon/policy.go:
```diff
@@ -252,11 +252,12 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *AddOptions, resCha
 	// taking the policy lock. As long as policyAdd is fed by a single-threaded
 	// queue this should never be an issue.
-	rules := d.dnsRuleGen.PrepareFQDNRules(sourceRules)
+	// TODO: ian fix me
```
What needs fixing here?

aanm replied:
+1

ianvernon replied:
Fixed with more descriptive comment - basically we are not pre-populating the selectorcache in this case. I think this is OK to defer for now.

joestringer commented on pkg/policy/cidr.go:
```diff
@@ -68,3 +68,26 @@ func GetCIDRPrefixes(rules api.Rules) []*net.IPNet {
+		return nil
+	}
+
+	res := make([]*net.IPNet, 0, 32)
```
`32` -> `len(ips)`?

joestringer commented on pkg/policy/cidr.go:
```diff
@@ -68,3 +68,26 @@ func GetCIDRPrefixes(rules api.Rules) []*net.IPNet {
+		prefix := &net.IPNet{
+			IP:   ip,
+			Mask: net.CIDRMask(bits, bits),
+		}
```
The above few lines look rather familiar (I see similar in `ParseCIDRs()`). Could be reused.

joestringer reviewed (changes requested):
I posted a few requests below about better sharing of code. I also noticed some chunks that should not be necessary.

Main bit of feedback is with ratelimiting of selectorcache update propagation. I think that's worth having a discussion in the relevant thread above.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -199,9 +244,167 @@ func (l *labelIdentitySelector) updateSelections() {
+}
+
+// updateSelections updates the immutable slice representation of the
+// cached selections after the cached selections have been changed.
```
you are mixing up function behavior with package behavior. remove `after the cached selections have been changed.`

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -199,9 +244,167 @@ func (l *labelIdentitySelector) updateSelections() {
+		selections[i] = nid
+		i++
+	}
+	// Sort the numeric identities so that the map iteration order
```
Why does it make testing easier? Is it really better sorting slices for every `updateSelections()` calls being made instead of looping over the `selections` slice in tests?

ianvernon replied:
This was the behavior added by @jrajahalme - I folded this functionality into a single type and then used composition for both `fqdnSelector` and `labelIdentitySelector`.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -199,9 +244,167 @@ func (l *labelIdentitySelector) updateSelections() {
+// policy rule as a string) to to the provided list of identities. If the contents
+// of the cachedSelections differ from those in the identities slice, all
+// users are notified.
+func (sc *SelectorCache) UpdateFQDNSelector(fqdnKey string, identities []identity.NumericIdentity) {
```
does the fdqnKey really needs to be a string? What are the implications of the number of memory allocations for this string for each policy recalculation? Why can't this be a `FQDNSelector` instead? Even in tests you are using `FQDNSelector` [1]

ianvernon replied:
Good idea, converting to `api.FQDNSelector`.

aanm replied:
@ianvernon What I meant here was using `FQDNSelector` directly the key of the map, not to pass the structure and they use a `.String()`

ianvernon replied:
The `SelectorCache` also uses the string representation of `api.EndpointSelector` as the key in the map, so we can't use `api.FQDNSelector` as the key directly. Do you want the key to be an interface type?

What about caching the contents of `String()` within a selector once we use it in the `SelectorCache`? This would make calls to `String()` only have an upfront cost of an initial computation, and since selectors are immutable, their string representation would always be the same. 

@jrajahalme do you have anything you'd like to add here?

jrajahalme replied:
`api.EndpointSelector` as a map key will be different for each `policy import` event if the selectors are the same. This is due to the `api.EndpointSelector` only containing pointers, which will be different for each re-creation of the same selector from JSON, etc. I haven't looked at the details of the `api.FQDNSelector`, but it could have the same issue.

Caching the string representation in the `api.Selector` looks like a sound proposal to me.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -199,9 +244,167 @@ func (l *labelIdentitySelector) updateSelections() {
+			log.WithFields(logrus.Fields{
+				"key":       fqdnKey,
+				"valueType": reflect.TypeOf(fqdnSel),
+			}).Error("key provided to UpdateFQDNSelector does not map to an FQDNSelector")
```
Why logging this as an error? How can the user fix it? As a user I don't even know what is the `UpdateFQDNSelector`

ianvernon replied:
This would be a problem on the programming side - I will change this to `Fatal` instead since this is something that should never, ever happen.

ianvernon replied:
Actually, given that I am changing the parameter for the function to be `api.FQDNSelector` instead of `string`, I can just remove this.

raybejjani commented on daemon/fqdn.go:
```diff
@@ -73,11 +76,88 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
+				if currentlyAllocatedIdentities, err = ipcache.AllocateCIDRs(bpfIPCache.IPCache, selectorIPNets); err != nil {
+					_ = d.prefixLengths.Delete(ipNets)
+					secIDCache.ReleaseSlice(context.Background(), usedIdentities)
+					log.WithError(err).WithField("prefixes", ipNets).Warn(
```
Should this be `context.TODO`? I imagine one day we will want a real timeout for it?

ianvernon replied:
Good catch, fixing.

aanm commented on pkg/policy/selectorcache_test.go:
```diff
@@ -246,6 +262,63 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdates(c *C) {
+	c.Assert(exists, Equals, true)
+
+	user1 := newUser(c, "user1", sc)
+	cached := user1.AddFQDNSelector(ciliumSel)
```
[1] even here you pass a `api.FQDNSelector` and not a `FQDNSelector.String()`.

aanm commented on pkg/policy/selectorcache_test.go:
```diff
@@ -246,6 +262,63 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdates(c *C) {
+
+	selections := cached.GetSelections()
+	c.Assert(len(selections), Equals, 3)
+	c.Assert(selections[0], Equals, identity.NumericIdentity(123))
```
You should check if both the selection slice contains all `ciliumIdentities`

aanm commented on pkg/policy/selectorcache_test.go:
```diff
@@ -246,6 +262,63 @@ func (ds *SelectorCacheTestSuite) TestIdentityUpdates(c *C) {
+	// Current selections contain the numeric identities of existing identities that match
+	selections2 := cached2.GetSelections()
+	c.Assert(len(selections2), Equals, 3)
+	c.Assert(selections2[0], Equals, identity.NumericIdentity(321))
```
You should check if both the selection slice contains all `googleIdentities`

aanm commented on pkg/fqdn/config.go:
```diff
@@ -45,9 +46,10 @@ type Config struct {
+	// as well as update the mapping of FQDNSelector to set of IPs.
 	// When set to nil, it is a no-op.
-	AddGeneratedRules func([]*api.Rule) error
+	AddGeneratedRulesAndUpdateSelectors func([]*api.Rule, map[string][]net.IP, []string) error
```
please add naming to the function arguments

aanm commented on pkg/fqdn/helpers.go:
```diff
@@ -57,14 +57,16 @@ func injectToCIDRSetRules(rule *api.Rule, cache *DNSCache, reMap *regexpmap.Rege
 
 		// Generate CIDR rules for each FQDN
 		for _, ToFQDN := range egressRule.ToFQDNs {
+			fqdnAsString := ToFQDN.String()
```
use `ToFQDN` directly.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -330,9 +532,17 @@ func AddIdentitySelector(user CachedSelectionUser, selector api.EndpointSelector
+	selectorCache.RemoveSelector(user, selector)
+}
+
+func UpdateFQDNSelector(fqdnKey string, identities []identity.NumericIdentity) {
```
Create a function that receives multiple FQDNKeys and a slice of identities to do [2]

ianvernon replied:
Can you clarify what the advantage of this is? Less locking on the SelectorCache?

aanm replied:
@ianvernon yes

ianvernon replied:
@aanm OK, will have a follow-up commit coming shortly once I get the build to pass without VM image downloads failing

aanm commented on daemon/fqdn.go:
```diff
@@ -73,11 +76,88 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
+				log.WithFields(logrus.Fields{
+					"fqdnSelectorString": selectorsWithoutIPs[k],
+				}).Debug("removing all identities from FQDN selector")
+				policy.UpdateFQDNSelector(selectorsWithoutIPs[k], []identity.NumericIdentity{})
```
With [2] done you can call `policy.UpdateFQDNSelector` directly for all `selectorsWithoutIPs` at the same time.

Commit:
ip: factor out common logic into helper functions

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
fqdn: plumb mapping of FQDNSelector --> set of IPs to SelectorCache

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -196,12 +170,182 @@ func (l *labelIdentitySelector) updateSelections() {
+//
+
+// lock must be held
+func (l *labelIdentitySelector) notifyUsers(added, deleted []identity.NumericIdentity) {
```
I think you can just drop this function since it should be composed from `selectorManager`.

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -315,12 +459,11 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
-}
-
-// Export global functions to interface with the global selector cache
+} // Export global functions to interface with the global selector cache
```
nit: not sure what this comment is supposed to relate to but seems it's moving up a bit, is it related to the function that the close brace is closing?

ianvernon replied:
Fixed the formatting.

joestringer commented on pkg/ipcache/cidr.go:
```diff
@@ -29,12 +29,40 @@ import (
+// any allocation fails, all allocations are rolled back and the error is
+// returned. When an identity is freshly allocated for a CIDR, it is added to
+// the ipcache.
+func AllocateCIDRsForIPs(prefixes []*net.IPNet) ([]*identity.Identity, error) {
```
Why does this take `[]*net.IPNet` and not `[]*net.IP`?

Just wondering if there's a way to avoid the prefix validation below since it seems like the caller is guaranteed to have real IPs and we should be able to programmatically generate full masks (`/32` for IPv4, `/128` for IPv6).

ianvernon replied:
This would indeed be simpler :) fixed!

joestringer reviewed (changes requested):
Some minor nits below. Would be good to get a review from @jrajahalme as well.

raybejjani commented on pkg/fqdn/helpers.go:
```diff
@@ -91,7 +93,7 @@ func injectToCIDRSetRules(rule *api.Rule, cache *DNSCache, reMap *regexpmap.Rege
 
 				// Mark this pattern missing; it will be unmarked in the loop below
-				missing[ToFQDN.MatchPattern] = struct{}{}
+				missing[ToFQDN] = struct{}{}
```
I think it's fine because we don't usually have matchName and matchPattern set in the same rule, but because we are using the whole selector we have to check that we didn't find IPs for the matchName in this selector and vice versa. :/

ianvernon replied:
Do you think we should require that MatchName and MatchPattern be a OneOf? We could always split them up into two separate `FQDNSelectors` in the repository for easier management? :/

raybejjani commented on pkg/policy/selectorcache.go:
```diff
@@ -196,12 +170,182 @@ func (l *labelIdentitySelector) updateSelections() {
+	*  Case 1: identities did correspond to this FQDN, but no longer do. Reset
+	*  the map
+	 */
+	if len(identities) == 0 && len(fqdnSel.cachedSelections) != 0 {
```
This case is basically the default case 3, right? since that would iterate over cachedSelection, deleting each one sequentially because idAsMap is empty. The range identities would also be a no-op.
This is more to make sure I understand it. I think the separate case is more clear.

ianvernon replied:
Correct!

raybejjani commented on pkg/policy/selectorcache.go:
```diff
@@ -315,7 +459,8 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
 		case *fqdnSelector:
-			// TODO
+			// This is a no-op right now. We don't encode in the identities
+			// which FQDNs they correspond to.
```
It may be problematic to try to include this anyway. The DNSName -> IP may be a many to one mapping, and it will change over time but the identity itself won't. This might also conflict with an IP when it is directly added via toCIDRs. The selector -> IP mapping might have a larger cardinality too, since `MatchPattern: "*"` is a common rule for L7, and could be for L3 too.

raybejjani reviewed (commented):
I'll need to look through this again; I feel I haven't fully thought through all of it. Generally, though, nothing looks problematic!

Commit:
policy: factor out mutually-exclusive portion of UpdateFQDNSelector into separate function

This allows the locking to be separate from updating a given selector if needed.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: provide functionality to remove identities from multiple FQDNSelectors

This allows the lock to be held once for multiple deletion operations.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
fqdn: utilize new function to remove IPs for set of FQDNSelector

Signed-off by: Ian Vernon <ian@cilium.io>

jrajahalme reviewed (approved):
I reviewed the selectorcache parts, the DNS proxy parts are too unfamiliar for me to review right now.
